File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int answer = 0;

void go(vector<int> numbers, int target,int sumNumber ,int idx){
    if(idx == numbers.size()){
        if (sumNumber == target){
            answer++;
            return;
        }
        else return;
    }
    
    go(numbers,target,sumNumber + numbers[idx], idx+1);
    go(numbers,target,sumNumber - numbers[idx], idx+1);
    
    return;
}

int solution(vector<int> numbers, int target) {
    go(numbers,target,0,0);
    int temp = answer;
    return temp;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int answer = 0;
    
    public void dfs(int idx, int size, int target,int num, int[] numbers){
        if (idx == size){
            if(num == target)answer++;
            return;
        }
        
        dfs(idx+1,size,target,num-numbers[idx],numbers);
        
        dfs(idx+1,size,target,num+numbers[idx],numbers);
    }
    
    public int solution(int[] numbers, int target) {
        
        dfs(0,numbers.length,target,0,numbers);
        return answer;
    }
}